### PR TITLE
Adding ml.net centos6 and ubuntu 16.04 container for openmp support

### PR DESCRIPTION
**Pull request by singlis**
- Updates the docker containers for centos and ubuntu to build clang with openmp support.

 Fixes issue #93.

### CODE REVIEW

**Comment by MichaelSimons** [comment]
@singlis - Have you considered added these ML specific components into a ML specific image that is built on top of the core images?  The primary benefit of this type of layering is it keeps the core image more manageable from a size perspective as well as tracking who depends on the various special components.

**Comment by shauheen** [comment]
Thanks @MichaelSimons for the suggestion. However OpenMP is not really ML specific. I agree with you completely that any specific requirement should be processed as you are suggesting.

**Comment by MichaelSimons** [comment]
@singlis - By ML specific I mean it is only being used by the ML product.  What other teams are going to be using OpenMP?

**Comment by eerhardt** [comment]
One issue with making a new docker image is the way we are building clang in the centos images. The new docker image would need to duplicate the clang build logic, add the `openmp` specific stuff into it, and re-make-install the whole clang again.

For the `ubuntu` images, it looks like it would be easy. @MichaelSimons, where would you suggest a new dockerfile be created?

**Comment by singlis** [comment]
Thank you everyone for the replies. 
@MichaelSimons  - yes it is specific to ML.net as I don't know of other teams that have a need for OpenMP. However it is a feature in Clang that I am enabling, so I tend to agree with Shauheen.

Eric also brings up a good issue with centos. @MichaelSimons - do you know if OpenMP can be side-loaded after clang is installed? I can try downloading only openmp and see if that can be compiled.

**Comment by scottqb** [comment]
I confirmed that for centos, I can build and install openmp without having to reinstall all of clang.

**Comment by singlis** [comment]
I confirmed on centos images that I am able to add OpenMP without having re-install Clang.

**Comment by singlis** [comment]
@MichaelSimons Let me know how you would like to proceed. I can create an image specifically for ML.net based on the centos 7 and ubuntu 16.04 images -- but where should they be placed? Should they go in this repository?

If you feel that this should be part of the dot net images, it maybe worth updating all images to include these changes so thats its a consistent change (its the main reason why I made this a WIP).

**Comment by MichaelSimons** [comment]
Since nobody is dependent on it outside of ML, my preference would be to put them in a ML specific image.  I am alright with including them in this repo.  Just put them in a directory underneath the os folder e.g. `src/centos/6/ML/Dockerfile`.  If there is a more generic name other that would apply other than ML that would be great.

File: src/centos/7/mlnet/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7
```
Comment (singlis) [inline src/centos/7/mlnet/Dockerfile]: @MichaelSimons - can you confirm if this is correct?
Also, how do I test this?

Comment (MichaelSimons) [reply src/centos/7/mlnet/Dockerfile]: You can test via `.\build.ps1 -DockerfilePath "centos/7*"`
Yes these should be part of the build.

Comment (MichaelSimons) [reply src/centos/7/mlnet/Dockerfile]: You will need to add these to the manifest.json at the root of the repo in order for them to get built.

Comment (MichaelSimons) [reply src/centos/7/mlnet/Dockerfile]: See [How to modify or create a new image](https://github.com/dotnet/dotnet-buildtools-prereqs-docker#how-to-modify-or-create-a-new-image) for details on this.

**Comment by singlis** [comment]
@MichaelSimons I have created ml.net containers for centos7 and ubuntu16.04. Can you review when you have a chance? Thank you very much for your help!

**Comment by singlis** [comment]
opensuse is failing due to the opensuse docker container 42.1 is no longer available. I can fix this in my PR if needed.

File: src/centos/7/mlnet/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+RUN yum install -y \
+        perl-Data-Dumper
+
+RUN wget http://releases.llvm.org/3.9.1/openmp-3.9.1.src.tar.xz && \
```
Comment (MichaelSimons) [inline src/centos/7/mlnet/Dockerfile]: A comment here would be valuable IMO just calling out the purpose of the layer - install OpenMP.  This was called out in the Ubuntu Dockerfile and that was more obvious why not here as well😸.

Comment (singlis) [reply src/centos/7/mlnet/Dockerfile]: Thanks Michael - since the opensuse issue causes the build to fail, its blocking this PR from going in.

Comment (singlis) [reply src/centos/7/mlnet/Dockerfile]: ...or maybe I dont have permissions to merge? I will update the PR per your comment.

Comment (MichaelSimons) [reply src/centos/7/mlnet/Dockerfile]: Not everyone has permission to merge.  I will merge once all legs are green except for opensuse.

**Comment by MichaelSimons** [comment]
FYI - the centos build leg is failing.  It is being investigated as time permits.  We are busy with release activities today.

The ubuntu build pass and produced the following images:

dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-207e097-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-debpkg-cfdd435-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-coredeps-d770b8b-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-crossdeps-cfdd435-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-cross-arm64-cfdd435-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-cross-arm64-alpine-cfdd435-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-cross-cfdd435-20190312152303
dotnetdocker.azurecr.io/public/dotnet-buildtools/prereqs:ubuntu-16.04-mlnet-207e097-20190312152303

**Comment by MichaelSimons** [comment]
The following centos/7 build issue was resolved.

```
mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-df8e152-20190312232803
mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-mlnet-df8e152-20190312232803
```

**Comment by singlis** [comment]
Thanks @MichaelSimons  - does it take some time for the docker images to become available? I updated my PR for ML.Net to reference the new docker containers, but it fails to find them. Here is the pull request for reference: 
https://github.com/dotnet/machinelearning/pull/2867

**Comment by MichaelSimons** [comment]
Your references are not correct, all new images from this repo are now being published to mcr.  Note the fully qualified tag names I mentioned earlier - e.g. `mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-mlnet-df8e152-20190312232803`

**Pull request merged**